inc/saitypes.h

```diff
@@ -1454,6 +1454,26 @@ typedef struct _sai_stat_capability_list_t
 
 } sai_stat_capability_list_t;
 
+/**
+ * @brief Bulk objects get statistics.
+ *
+ * @param[in] object_count Number of objects to get on attribute
+ * @param[in] object_id List of object ids
+ * @param[in] number_of_counters Number of counters in the array
+ * @param[in] counter_ids Specifies the array of counter ids
+ * @param[in] mode Statistics mode
+ * @param[out] counters Array of resulting counter values
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_bulk_object_get_stats_fn)(
+        _In_ uint32_t object_count,
+        _In_ const sai_object_id_t *object_id,
+        _In_ uint32_t number_of_counters,
```

Comment:
this api assumes that you query same number of counters for each object, is that what it should be ? should it be more like bulk_get ?

Reply:
> 

yes, api assumes number of counters queried are same for all the objects in the list.

Reply:
I think it should be more similar to bulk get then

sai_status_t sai_bulk_get_attribute(
        _In_ sai_object_id_t switch_id,
        _In_ sai_object_type_t object_type,
        _In_ uint32_t object_count,
        _In_ const sai_object_key_t *object_key,
        _Inout_ uint32_t *attr_count,
        _Inout_ sai_attribute_t **attr_list,
        _Inout_ sai_status_t *object_statuses);

1) it should have switch id
2) it should have object type, to make clear that all objects are from the same type, and identify the stats correctly, as the stats ranges overlap same as attributes. We can identify the object type from the object id, but since we give the explicit type for bulk get, i think it should also be given here to make things clear.
3) might be better to work with object key and not object id, in case we ever want to query a counter for an object_key that doesn't have object id
4) i think object statuses is also needed

otherwise, i think it's a good initiative, that can help to optimize counter reading

Reply:
btw, we can add a new function under each api which has stats
or have one general function in saiobject.h like sai_bulk_get_attribute
i'm fine with both approaches

Reply:
> 

I explored that options. It is more intrusive change and lot of tooling code was getting modified. Though back ward compatibility was not an issue as it would be a NULL vector for older implementations.

I feel current proposal is simpler and given that it is now enhanced with object type, it is explicit in its intent with single API for all the object types.

Reply:
> 

Looks like saitypes.h do not include saiobject.h (circular dependency), so we may not be able to specify the object as key. We can always get the object key from object id to query, right?

Reply:
https://github.com/opencomputeproject/SAI/pull/1352/commits/96ecc1c4407d8af776bddd862404b28a755c3f54

Reply:
i thought we can add the new function just after
sai_bulk_get_attribute
in saiobject.h

That way it can have object_key defined

Reply:
> 

we need to modify styles.pm and few more places where API name validations are performed.

Reply:
> i thought we can add the new function just after sai_bulk_get_attribute in saiobject.h
> 
> That way it can have object_key defined

Yes, we can do this

Reply:
> > i thought we can add the new function just after sai_bulk_get_attribute in saiobject.h
> > That way it can have object_key defined
> 
> Yes, we can do this

https://github.com/opencomputeproject/SAI/pull/1352/commits/96ecc1c4407d8af776bddd862404b28a755c3f54